### PR TITLE
Improvement:added -f --file argument for .yaml .yml

### PR DESCRIPTION
**Pull request by siddhant-basepair**
User should be able to create and update Module or Pipeline via -f or —file
Example -
create module —file1 test.yaml or test.yml
create moudle -f test.yaml or test.yml

Related Ticket -
https://app.asana.com/0/1201368373933132/1201589980257997

### CODE REVIEW

File: bin/basepair [hunk 1]
```diff
@@ -523,10 +523,10 @@ def add_tags_parser(parser):
   return parser
 
 def add_yaml_parser(parser):
-  '''Add uid parser'''
+  '''Add yaml file parser'''
   parser.add_argument(
-    '-y',
-    '--yaml',
+    '-f',
+    '--file1',
```
Comment (Lalee10) [inline bin/basepair:529]: What issue is faced when using `--file`?

**Review by Lalee10 — CHANGES_REQUESTED** [review]

File: bin/basepair (hunk 1 above)
Comment (siddhant-basepair) [reply bin/basepair]: changed it to file as well

File: basepair/api.py [hunk 2]
```diff
@@ -434,20 +434,29 @@ def create_module(self, data):
...
+            self.update_module(data)
+            return None
     module_id = info.get('id')
     module_name = info.get('name')
     if module_id:  # success
       if self.verbose:
         eprint(f'created: module {module_name} with id {module_id}')
     else:  # failure
-      eprint('failed module creation:', info.get('error'))
+      eprint('failed creating module')
```
Comment (Lalee10) [inline basepair/api.py:454]: log some information on why module creation failed, can we reuse `info.get('error')` here?

File: basepair/api.py [hunk 3]
```diff
@@ -466,7 +475,7 @@ def update_module(self,data):
         module_name = info.get('name')
         eprint(f'updated: module {module_name} with id {module_id}')
     else:  # failure
-      eprint('failed module creation:', info.get('error'))
+      eprint('failed updating module')
```
Comment (Lalee10) [inline basepair/api.py:478]: log some information on why module **update** failed, can we reuse `info.get('error')` here?

File: basepair/api.py [hunk 4]
```diff
@@ -501,13 +510,24 @@ def create_pipeline(self,data):
...
+            self.update_pipeline(data)
+            return None
     workflow_id = info.get('id')
     workflow_name = info.get('name')
     if workflow_id:  # success
       if self.verbose:
         eprint(f'created: workflow {workflow_name} with id {workflow_id}')
     else:  # failure
-      eprint('failed workflow creation:',info.get('error'))
+      eprint('failed creating pipeline')
```
Comment (Lalee10) [inline basepair/api.py:530]: log some information on why pipeline creation failed, can we reuse `info.get('error')` here?

File: basepair/api.py [hunk 5]
```diff
@@ -531,7 +551,7 @@ def update_pipeline(self,data):
         workflow_name = info.get('name')
         eprint(f'updated: workflow {workflow_name} with id {workflow_id}')
     else:  # failure
-      eprint('failed workflow update:', info.get('error'))
+      eprint('failed updating pipeline')
```
Comment (Lalee10) [inline basepair/api.py:554]: log some information on why pipeline update failed, can we reuse `info.get('error')` here?

File: basepair/api.py [hunk 6]
```diff
@@ -1501,6 +1521,33 @@ def _get_user_id(self):
     )
     self.conf['user'] = user.get('objects', [None])[0]
 
+  def yes_or_no(self,question):
+    valid = {
+      'yes': True,
+      'y': True,
+      'ye': True,
```
Comment (Lalee10) [inline basepair/api.py:1528]: remove `ye` as an acceptable answer, rest are fine

File: basepair/api.py [hunk 7]
```diff
@@ -1501,6 +1521,33 @@ def _get_user_id(self):
...
+    }
+
+    prompt = ' [y/n] '
+
+    while True:
+
+      # get input function for py 2 and 3
+      get_input = input
+      if sys.version_info[:2] <= (2, 7):
+        get_input = raw_input
```
Comment (Lalee10) [inline basepair/api.py:1540]: We don't support python 2, so you can just add code for python 3

image

**Review by Lalee10 — CHANGES_REQUESTED** [review]

File: basepair/api.py (hunk 2 above)
Comment (siddhant-basepair) [reply basepair/api.py:454]: error is already logging once without it so updated the over all status failure

File: basepair/api.py (hunk 3 above)
Comment (siddhant-basepair) [reply basepair/api.py:478]: API error's is already logging once without it so updated the overall status failure

File: basepair/api.py (hunk 4 above)
Comment (siddhant-basepair) [reply basepair/api.py:530]: API error's is already logging once without it so updated the overall status failure

File: basepair/api.py (hunk 5 above)
Comment (siddhant-basepair) [reply basepair/api.py:554]: API error's is already logging once without it so updated the overall status failure

File: basepair/api.py (hunk 6 above)
Comment (siddhant-basepair) [reply basepair/api.py:1528]: done

File: basepair/api.py (hunk 7 above)
Comment (siddhant-basepair) [reply basepair/api.py:1540]: Done

File: basepair/api.py [hunk 8]
```diff
@@ -1501,6 +1521,30 @@ def _get_user_id(self):
...
+      'no': False,
+      'n': False
+    }
+
+    prompt = ' [y/n] '
+
+    while True:
+
+      # get input function for py 2 and 3
+      get_input = input
```
Comment (Lalee10) [inline basepair/api.py:1537]: remove unnecessary variable declaration and use directly

File: basepair/api.py (hunk 2 above)
Comment (Lalee10) [reply basepair/api.py:454]: standardize the message `failed module creation`

File: basepair/api.py (hunk 3 above)
Comment (Lalee10) [reply basepair/api.py:478]: change message -> `failed module update`

File: basepair/api.py (hunk 4 above)
Comment (Lalee10) [reply basepair/api.py:530]: change message to `failed pipeline creation`

File: basepair/api.py (hunk 5 above)
Comment (Lalee10) [reply basepair/api.py:554]: change message to `failed pipeline creation`

File: bin/basepair [hunk 9]
```diff
@@ -393,29 +393,29 @@ def update_analysis(bp, analysis_id, keys, vals):
...
     eprint('ERROR: Yaml file required.')
     return
-  if len(args.yaml) != 1:
+  if len(args.file) != 1:
     eprint('Please provide only one yaml')
     return
-  yaml_path = args.yaml[0]
-  if args.yaml[0].endswith('.yaml'):
+  yaml_path = args.file[0]
+  if args.file[0].endswith('.yaml') or args.file[0].endswith('.yml'):
```
Comment (Lalee10) [inline bin/basepair:403]: use declared variable `yaml_path`, you can also simplify the condition by using a tuple as the suffix for `endsWith`

```Python
if yaml_path.endsWith('yaml', 'yml'):
```

File: bin/basepair [hunk 10]
```diff
@@ -393,29 +393,29 @@ def update_analysis(bp, analysis_id, keys, vals):
...
```
Comment (Lalee10) [inline bin/basepair:418]: use declared variable `yaml_path`, you can also simplify the condition by using a tuple as the suffix for `endsWith`

```Python
if yaml_path.endsWith('yaml', 'yml'):
```

File: bin/basepair [hunk 11]
```diff
@@ -228,30 +228,30 @@ def create_analysis(bp, args):
...
     return
-  if len(args.yaml) != 1:
+  if len(args.file) != 1:
     eprint('Please provide only one yaml')
     return
-  yaml_path = args.yaml[0]
-  if args.yaml[0].endswith('.yaml'):
-    bp.create_module({'yamlpath': yaml_path})
+  yaml_path = args.file[0]
+  if args.file[0].endswith('.yaml') or args.file[0].endswith('.yml'):
```
Comment (Lalee10) [inline bin/basepair:238]: use declared variable `yaml_path`, you can also simplify the condition by using a tuple as the suffix for `endsWith`

```Python
if yaml_path.endsWith('yaml', 'yml'):
```

File: bin/basepair [hunk 12]
```diff
@@ -228,30 +228,30 @@ def create_analysis(bp, args):
...
     return
-  if len(args.yaml) != 1:
+  if len(args.file) != 1:
     eprint('Please provide only one yaml')
     return
-  yaml_path = args.yaml[0]
-  if args.yaml[0].endswith('.yaml'):
-    bp.create_pipeline({'yamlpath': yaml_path})
+  yaml_path = args.file[0]
+  if args.file[0].endswith('.yaml') or args.file[0].endswith('.yml'):
```
Comment (Lalee10) [inline bin/basepair:253]: use declared variable `yaml_path`, you can also simplify the condition by using a tuple as the suffix for `endsWith`

```Python
if yaml_path.endsWith('yaml', 'yml'):
```

**Review by Lalee10 — CHANGES_REQUESTED** [review]
In each create and update method (both pipeline and module) we are repeating a number of step i.e.

1. Check if file provided
2. Check if only single file provided
3. Check if file YAML file

Please move this code to a common function

**Review by Lalee10 — APPROVED** [review]

**Pull request merged**